### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by dhrubo-os**
### Description
[initial commit]
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: opensearch_py_ml/ml_models/metrics_correlation/utils.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+# Any modifications Copyright OpenSearch Contributors. See
+# GitHub history for details.
+
+BANDWIDTH = 16.0
```
Comment (greaa-aws) [inline opensearch_py_ml/ml_models/metrics_correlation/utils.py:8]: Should make this data-dependent, specifically a function of $T$.

File: opensearch_py_ml/ml_models/metrics_correlation/dip.py [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+# GitHub history for details.
+
+from typing import Tuple
+
+import torch
+
+
+def dip(x: torch.Tensor) -> float:
+    """
+    Fast computation of the dip test statistic. See: https://www.jstor.org/stable/pdf/2241144.pdf (paper)
```
Comment (greaa-aws) [inline opensearch_py_ml/ml_models/metrics_correlation/dip.py:15]: Both functions here need to be documented a little better. I can help with that.

File: opensearch_py_ml/ml_models/metrics_correlation/dip.py [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    :param x:
+    :type x: torch.Tensor
+    :return:
+    :rtype: Tuple[float, float]
+    """
+
+    # torchScript doesn't support tensor values as global variable so we can't put these
+    # constant values in utils.py
+
+    N_VALS: torch.Tensor = torch.tensor(
```
Comment (greaa-aws) [inline opensearch_py_ml/ml_models/metrics_correlation/dip.py:210]: The p-value lookup table here has a sample size (i.e. length of time series) range of $T=100$ to $T=50000$. The upper bound is probably fine but we should consider whether the lower bound is too high. The method doesn't break if the input has a length outside of these bounds, but the p-value calculation and thus event selection can be inaccurate. We should consider whether a wider range needs to be supported.

File: opensearch_py_ml/ml_models/metrics_correlation/mcorr.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        return events
+
+    def compute_activity_scores(
+        self, metrics: torch.Tensor, max_events: int, wavelet_approx: bool
+    ) -> torch.Tensor:
+        M, T = metrics.shape
+        activity_scores = torch.zeros((M, T - 1))
+        for m in range(
+            M
+        ):  # NOTE: ideally we'd be able to parallelize this, as in the non-script version
```
Comment (greaa-aws) [inline opensearch_py_ml/ml_models/metrics_correlation/mcorr.py:76]: Even after wavelet acceleration, the main computational bottleneck of this method is the per-metric activity scoring. Parallelization can help this algorithm scale to even larger data. It could be a good topic for future investigation.

File: opensearch_py_ml/ml_models/metrics_correlation/nmf.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        :type W: torch.Tensor
+        :param H: H is a data metrics of dimensions K × N with non-negative entries
+        :type H: torch.Tensor
+        :return:
+        :rtype: Tuple[torch.Tensor, torch.Tensor]
+        """
+        # W update
+        VHt = torch.mm(V, H.t())
+        WHHt = torch.mm(W, torch.mm(H, H.t()))
+        # WHHt[WHHt==0] = torch.finfo(WHHt.dtype).eps # for safe division
```
Comment (greaa-aws) [inline opensearch_py_ml/ml_models/metrics_correlation/nmf.py:41]: Can delete this commented line and the one below.

File: opensearch_py_ml/ml_models/metrics_correlation/nmf.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        """
+        U, S, Vh = torch.linalg.svd(V, full_matrices=False)
+
+        d = U.shape[1]  # factors are identified up to sign. make sure they're positive
+        for dd in range(d):
+            if torch.any(U[:, dd] < 0):
+                U[:, dd] *= -1
+                Vh[dd, :] *= -1
+
+        k = int(k)  # not sure why this is needed but torchscript breaks without it
```
Comment (greaa-aws) [inline opensearch_py_ml/ml_models/metrics_correlation/nmf.py:77]: Should probably clean stuff like this up.

File: opensearch_py_ml/ml_models/metrics_correlation/nmf.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        for i in range(1, max_iter + 1):
+            W, H = self.step(V, W, H)
+
+            if i % 10 == 0:  # check rel err stopping criterion
+                err = torch.linalg.norm(V - torch.mm(W, H))
+                if (prev_err - err) / init_err < tol:
+                    break
+                prev_err = err
+
+            if i == max_iter:
```
Comment (greaa-aws) [inline opensearch_py_ml/ml_models/metrics_correlation/nmf.py:141]: Need a different way to report convergence failures.

File: opensearch_py_ml/ml_models/metrics_correlation/wavelet_tools.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    from a truncated Haar wavelet representation.
+"""
+
+import math
+
+# imports
+import torch
+import torch.nn.functional as F
+
+# hardcode haar wavelet and scale filters
```
Comment (greaa-aws) [inline opensearch_py_ml/ml_models/metrics_correlation/wavelet_tools.py:31]: Remove old commented global vars.

File: opensearch_py_ml/ml_models/metrics_correlation/wavelet_tools.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+# HAAR_HI = torch.FloatTensor([-1/math.sqrt(2), 1/math.sqrt(2)])
+# HAAR_LO = torch.FloatTensor([1/math.sqrt(2), 1/math.sqrt(2)])
+
+
+def haar_approx(signal: torch.Tensor, truncate: int) -> torch.Tensor:
```
Comment (greaa-aws) [inline opensearch_py_ml/ml_models/metrics_correlation/wavelet_tools.py:36]: Leaving in here as a pointer to future work: there are opportunities to further optimize these wavelet tools, which in turn might help make them more broadly usable. Ex: DFT vs wavelet matrix approach to computing the DWT, batch wavelet approximation of multiple series in parallel.

File: tests/ml_models/test_mcorr_pytest.py [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+# Any modifications Copyright OpenSearch Contributors. See
+# GitHub history for details.
+
+
+import torch
+
+from opensearch_py_ml.ml_models import MCorr
+
+
+def test_mcorr_output():
```
Comment (greaa-aws) [inline tests/ml_models/test_mcorr_pytest.py:14]: Need to add more tests for this module. In addition to end-to-end performance, we probably want tests for individual components to localize errors or regressions.

File: opensearch_py_ml/ml_models/metrics_correlation/nmf.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+# compatible open source license.
+# Any modifications Copyright OpenSearch Contributors. See
+# GitHub history for details.
+
+from typing import Tuple
+
+import torch
+
+
+class NMF(torch.nn.Module):
```
Comment (greaa-aws) [inline opensearch_py_ml/ml_models/metrics_correlation/nmf.py:13]: We previously discussed whether this should be a `nn.Module` on its own vs. just a function called during `MCorr.forward()`. What's the current thinking on that?

File: opensearch_py_ml/ml_models/metrics_correlation/utils.py (hunk 1 above)
Comment (dhrubo-os) [reply opensearch_py_ml/ml_models/metrics_correlation/utils.py:8]: Yeah I added `self.bandwidth = float(max(T // 32, 1))` in the forward function as we discussed. I put this value for now as a initialization of this variable.

File: opensearch_py_ml/ml_models/metrics_correlation/dip.py (hunk 2 above)
Comment (dhrubo-os) [reply opensearch_py_ml/ml_models/metrics_correlation/dip.py:15]: Yeah, I left few functions, parameters of the functions without comments where I'm hoping your help. The target of this PR is just to setup the skeleton so that we can start gradually fleshing it out.

File: tests/ml_models/test_mcorr_pytest.py (hunk 10 above)
Comment (dhrubo-os) [reply tests/ml_models/test_mcorr_pytest.py:14]: Agree

**Review by greaa-aws — APPROVED** [review]

File: opensearch_py_ml/ml_models/metrics_correlation/nmf.py (hunk 11 above)
Comment (dhrubo-os) [reply opensearch_py_ml/ml_models/metrics_correlation/nmf.py:13]: Yeah, for now I'm keeping this way.
